BookingProject/BookingWebAPI/Migrator/Services/DataMigrationService.cs

```diff
@@ -0,0 +1,112 @@
+        {
+            _context = context;
+            _userManager = userManager;
+            _mapper = mapper;
+            _userRepository = userRepository;
+        }
+
+        public async Task<Result<bool>> MigrateData(string fileName)
+        {
+            string basePath = AppContext.BaseDirectory; 
+            string companyFilePath = Path.Combine(basePath, _dataDirectory, fileName);
+
+            if (!File.Exists(companyFilePath))
+            {
+                return Result<bool>.Failure("File not found");
+            }
+            
+            await using var dbTransaction = await _context.Database.BeginTransactionAsync();
+
+            try
+            {
+                var result = await MigrateFile(companyFilePath);
+
+                if(!result.IsSuccess)
+                {
+                    throw new Exception(result.ErrorMessage);
+                }
+
+                await dbTransaction.CommitAsync();
+                return Result<bool>.Success(true);
+            }
+            catch (Exception ex)
+            {
+                await dbTransaction.RollbackAsync();
+                return Result<bool>.Failure(ex.Message);
+            }
+        }
+
+        private async Task<Result<bool>> MigrateFile(string companyFilePath)
+        {
+            var hosts = await ReadDataFromFile<List<AppUserMigrationDto>>(companyFilePath);
+
+            if (hosts == null)
+            {
+                return Result<bool>.Failure("File not found");
+            }
+
+            foreach (var item in hosts)
+            {
+                if (await _userRepository.UserExistsByIdAndCompany(item.Id, item.SourceCompanyId))
+                {
+                    Console.WriteLine($"User with ExternalId: {item.Id} and SourceCompanyId: {item.SourceCompanyId} already exists. Skipping creation.");
+                    continue;
+                }
+
+                var newHost = _mapper.Map<AppUser>(item);
+                
+                var createdUser = await _userManager.CreateAsync(newHost,GenerateDefaultPassword());
+                if (!createdUser.Succeeded)
+                {
+                    var errors = createdUser.Errors.Select(x => x.Description).ToString();
+                    return Result<bool>.Failure(errors);
+                }
+                
+                var roleResult = await _userManager.AddToRoleAsync(newHost, UserRoles.Host);
+                if (!roleResult.Succeeded)
+                {
+                    var errorsRole = roleResult.Errors.Select(x => x.Description).ToList();
+                    return Result<bool>.Failure(errorsRole);
+                }
+            }
+            return Result<bool>.Success(true);
+        }
+        
+        private async Task<T?> ReadDataFromFile<T>(string filePath)
+        {
+            var data = await File.ReadAllTextAsync(filePath);
+            return JsonSerializer.Deserialize<T>(data);
+        }
```

Comment:
I dislike that we loading 5gb into memory. What if files get bigger? I suggest using stream deserialization
https://chatgpt.com/share/67349844-5258-800b-895d-3afc9cc12aa6

Reply:
Fixed